### PR TITLE
Introduce command runner

### PR DESCRIPTION
**Pull request by bbrto21**
* Enhance github actions
* Use python3

Signed-off-by: Boram Bae <boram21.bae@samsung.com>

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -7,37 +7,19 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v2
-      - uses: jitterbit/get-changed-files@v1
-        id: files
         with:
-          format: space-delimited
-          token: ${{ secrets.GITHUB_TOKEN }}
+          ref: ${{ github.event.pull_request.head.sha }}
```
Comment (WonyoungChoi) [inline .github/workflows/build.yml:11]: What happens to the push event?

File: tools/commands/check_tidy.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
```
Comment (bwikbs) [inline tools/commands/check_tidy.py:1]: Do you have any plans to backport it to original place? 😄

**Review by bwikbs — APPROVED** [review]
Overall, I love this enhancement. 👍 
If possible, I would like help message on the command line to be more useful, and other developers can use this by documentation. 😄

File: .github/workflows/build.yml (hunk 1 above)
Comment (bbrto21) [reply .github/workflows/build.yml]: I updated PR with using specific SHA, It works on both push and pull_request events now

File: tools/commands/check_tidy.py (hunk 2 above)
Comment (bbrto21) [reply tools/commands/check_tidy.py:1]: I don't get you... could you tell me more detail?

File: tools/commands/build_example.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import commands.command_utils as command_utils
+
+
+def parse_args(args):
+    parser = command_utils.get_options_parser(plugins=True, exclude=True, run_on_changed_packages=True, base_sha=True)
+    return parser.parse_args(args)
+
+
+def _build_examples(plugin):
+    print(f'============= Building for {plugin} =============')
```
Comment (HakkyuKim) [inline tools/commands/build_example.py:16]: For readability, `{plugin}` should display the basename only.

File: tools/commands/build_example.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                   stdout=subprocess.PIPE)
+
+    completed_process = subprocess.run('flutter-tizen build tpk --device-profile wearable -v',
+                                       shell=True,
+                                       cwd=example_dir,
+                                       universal_newlines=True,
+                                       stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+    print(completed_process.stdout)
+    if completed_process.returncode == 0:
+        print(f'============= Succeed to build {plugin} =============')
```
Comment (HakkyuKim) [inline tools/commands/build_example.py:31]: For readability, `{plugin}` should display the basename only.

File: tools/commands/build_example.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        return True
+    else:
+        print(f'============= Failed to build {plugin} =============')
```
Comment (HakkyuKim) [inline tools/commands/build_example.py:34]: For readability, `{plugin}` should display the basename only.

File: tools/commands/build_example.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        return False
+
+
+def run_build_examples(argv):
+    args = parse_args(argv)
+    packages_dir = command_utils.get_package_dir()
+    target_plugins, exclude_plugins = command_utils.get_target_plugins(
```
Comment (HakkyuKim) [inline tools/commands/build_example.py:41]: `exclude_plugins` is not used.

```suggestion
    target_plugins, _ = command_utils.get_target_plugins(
```

File: tools/commands/build_example.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+                   cwd=example_dir,
+                   stderr=subprocess.PIPE,
+                   stdout=subprocess.PIPE)
+
+    completed_process = subprocess.run('flutter-tizen build tpk --device-profile wearable -v',
+                                       shell=True,
+                                       cwd=example_dir,
+                                       universal_newlines=True,
+                                       stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+    print(completed_process.stdout)
```
Comment (HakkyuKim) [inline tools/commands/build_example.py:29]: Removing `subprocess.PIPE` shows logs in CI terminal in realtime.

```suggestion
    completed_process = subprocess.run('flutter-tizen build tpk --device-profile wearable -v',
                                       shell=True,
                                       cwd=example_dir)
```

File: tools/commands/build_example.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+
+def _build_examples(plugin):
+    print(f'============= Building for {plugin} =============')
+    example_dir = os.path.join(plugin, 'example')
+    subprocess.run('flutter-tizen pub get',
+                   shell=True,
+                   cwd=example_dir,
+                   stderr=subprocess.PIPE,
+                   stdout=subprocess.PIPE)
```
Comment (HakkyuKim) [inline tools/commands/build_example.py:22]: Removing `subprocess.PIPE` shows logs in CI terminal.

```suggestion
    subprocess.run('flutter-tizen pub get',
                   shell=True,
                   cwd=example_dir)
```

File: .github/workflows/build.yml [hunk 9]
```diff
@@ -7,37 +7,18 @@ jobs:
...
-          done
+          git fetch origin master
+          CHANGED_PACKAGES=$(python3 tools/run_command.py plugins --run-on-changed-packages --base-sha $(git rev-parse HEAD^) 2>&1)
+          echo $CHANGED_PACKAGES
           if [[ ${#CHANGED_PACKAGES[@]} -ne 0 ]]; then
               echo "CHANGED_PACKAGES=${CHANGED_PACKAGES[@]}" >> "$GITHUB_ENV"
           fi
       - name: Install prerequisite packages
-        if: env.CHANGED_PACKAGES != ''
+        if: env.CHANGED_PACKAGES != '[]'
```
Comment (HakkyuKim) [inline .github/workflows/build.yml:21]: If you print plugins in lines:

```suggestion
        if: env.CHANGED_PACKAGES != ''
```

File: .github/workflows/build.yml [hunk 10]
```diff
@@ -62,33 +43,28 @@ jobs:
           libxcb-xfixes0 \
           libxi6
       - name: Install Tizen Studio
-        if: env.CHANGED_PACKAGES != ''
+        if: env.CHANGED_PACKAGES != '[]'
```
Comment (HakkyuKim) [inline .github/workflows/build.yml:46]: Same here.

File: .github/workflows/build.yml [hunk 11]
```diff
@@ -62,33 +43,28 @@ jobs:
...
         run: |
           sudo apt install -y pciutils zip libncurses5 python libpython2.7
           curl http://download.tizen.org/sdk/Installer/tizen-studio_4.1/web-cli_Tizen_Studio_4.1_ubuntu-64.bin -o install.bin
           chmod a+x install.bin
           ./install.bin --accept-license $HOME/tizen-studio
           rm install.bin
       - name: Install Tizen Studio packages
-        if: env.CHANGED_PACKAGES != ''
+        if: env.CHANGED_PACKAGES != '[]'
```
Comment (HakkyuKim) [inline .github/workflows/build.yml:54]: Same here.

File: .github/workflows/build.yml [hunk 12]
```diff
@@ -62,33 +43,28 @@ jobs:
...
         run: |
           $HOME/tizen-studio/package-manager/package-manager-cli.bin install \
           NativeCLI \
           NativeToolchain-Gcc-9.2 \
           WEARABLE-4.0-NativeAppDevelopment \
           WEARABLE-5.5-NativeAppDevelopment
       - name: Install flutter-tizen
-        if: env.CHANGED_PACKAGES != ''
+        if: env.CHANGED_PACKAGES != '[]'
```
Comment (HakkyuKim) [inline .github/workflows/build.yml:62]: Same here.

File: .github/workflows/build.yml [hunk 13]
```diff
@@ -62,33 +43,28 @@ jobs:
...
         run: |
           git clone https://github.com/flutter-tizen/flutter-tizen.git
       - name: Build examples of changed packages
-        if: env.CHANGED_PACKAGES != ''
+        if: env.CHANGED_PACKAGES != '[]'
```
Comment (HakkyuKim) [inline .github/workflows/build.yml:66]: Same here.

File: tools/commands/build_example.py [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+        results.append(result)
+
+    if False not in results:
+        exit(0)
+    else:
+        exit(1)
+
+
+if __name__ == '__main__':
+    run_build_examples(sys.argv)
```
Comment (HakkyuKim) [inline tools/commands/build_example.py:56]: If `run_command.py` is now the entry point, we can remove module execution and shebang on top (Also applies to `check_tidy.py`, `integration_test.py`, and `plugins.py`).

File: tools/commands/plugins.py [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+
+import commands.command_utils as command_utils
+
+
+def parse_args(args):
+    parser = command_utils.get_options_parser(run_on_changed_packages=True, base_sha=True)
+    return parser.parse_args(args)
+
+
+def run_plugins(argv):
```
Comment (HakkyuKim) [inline tools/commands/plugins.py:13]:
```suggestion
def print_plugins(argv):
```

File: tools/run_command.py [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+
+
+# Build example app
+def run_build_examples(argv):
+    build_example.run_build_examples(argv)
+
+
+# Print plugin list
+def run_plugins(arv):
+    plugins.run_plugins(arv)
```
Comment (HakkyuKim) [inline tools/run_command.py:28]:
```suggestion
def print_plugins(arv):
    plugins.print_plugins(arv)
```

File: tools/run_command.py [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+
+
+commands = {}
+commands["tidy"] = run_check_tidy
+commands["test"] = run_integration_test
+commands["build"] = run_build_examples
+commands["plugins"] = run_plugins
```
Comment (HakkyuKim) [inline tools/run_command.py:35]:
```suggestion
commands["plugins"] = print_plugins
```

File: .github/workflows/build.yml [hunk 18]
```diff
@@ -7,37 +7,18 @@ jobs:
...
-          for changed_file in ${{ steps.files.outputs.all }}; do
-            if [[ "$changed_file" == "packages"*".cc" ]] || [[ "$changed_file" == "packages"*".h" ]]; then
-                package=$(echo ${changed_file} | sed 's|\/tizen.*||;s|.*packages\/||')
-                if ! list_include_item $CHANGED_PACKAGES $package; then
-                    CHANGED_PACKAGES+=($package)
-                fi
-            fi
-          done
+          git fetch origin master
+          CHANGED_PACKAGES=$(python3 tools/run_command.py plugins --run-on-changed-packages --base-sha $(git rev-parse HEAD^) 2>&1)
```
Comment (HakkyuKim) [inline .github/workflows/build.yml:15]: Question: for `pull_request` events does `HEAD^` return the tip of the target branch?

Comment (HakkyuKim) [reply .github/workflows/build.yml:15]: I've confirmed that it does.

File: tools/run_command.py [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+
+if __name__ == "__main__":
+    try:
+        commands[sys.argv[1]](sys.argv[2:])
+    except Exception as e:
+        print(e)
```
[the remark this answers is not in the document]
Comment (bbrto21) [reply tools/run_command.py]: Thank you for review~  I updated pr to enhance printing information to use this script

File: tools/commands/plugins.py [hunk 20]
```diff
@@ -0,0 +1,22 @@
...
+    args = parse_args(argv)
+    packages_dir = command_utils.get_package_dir()
+    target_plugins, exclude_plugins = command_utils.get_target_plugins(
```
[the remark this answers is not in the document]
Comment (bbrto21) [reply tools/commands/plugins.py]: I applied it! thanks

File: tools/commands/plugins.py [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+        packages_dir, run_on_changed_packages=args.run_on_changed_packages, base_sha=args.base_sha)
+    print(target_plugins)
```
[the remark this answers is not in the document]
Comment (bbrto21) [reply tools/commands/plugins.py]: I applied it! thanks

File: tools/commands/integration_test.py [hunk 22]
```diff
@@ -13,12 +14,16 @@
...
+
 import argparse
+import os
+import re
 import subprocess
 import sys
-import os
 import time
-import re
+from typing import Tuple
```
[the remark this answers is not in the document]
Comment (bbrto21) [reply tools/commands/integration_test.py]: done!

File: tools/commands/build_example.py (hunk 3 above)
Comment (bbrto21) [reply tools/commands/build_example.py]: I fixed it as your opinion

File: tools/commands/build_example.py (hunk 4 above)
Comment (bbrto21) [reply tools/commands/build_example.py]: ditto

File: tools/commands/build_example.py (hunk 5 above)
Comment (bbrto21) [reply tools/commands/build_example.py]: ditto

File: tools/commands/build_example.py (hunk 6 above)
Comment (bbrto21) [reply tools/commands/build_example.py]: thanks!

File: tools/commands/build_example.py (hunk 8 above)
Comment (bbrto21) [reply tools/commands/build_example.py]: Thanks for let me know!

File: .github/workflows/build.yml (hunk 9 above)
Comment (bbrto21) [reply .github/workflows/build.yml]: I revert this change!

File: .github/workflows/build.yml (hunk 11 above)
Comment (bbrto21) [reply .github/workflows/build.yml]: ditto

File: .github/workflows/build.yml (hunk 10 above)
Comment (bbrto21) [reply .github/workflows/build.yml]: ditto

File: .github/workflows/build.yml (hunk 12 above)
Comment (bbrto21) [reply .github/workflows/build.yml]: ditto

File: .github/workflows/build.yml (hunk 13 above)
Comment (bbrto21) [reply .github/workflows/build.yml]: ditto

File: tools/commands/build_example.py (hunk 14 above)
Comment (bbrto21) [reply tools/commands/build_example.py:54]: I wanted each file to be able to run independently too

File: tools/commands/plugins.py (hunk 15 above)
Comment (bbrto21) [reply tools/commands/plugins.py]: I renamed this

File: tools/run_command.py (hunk 16 above)
Comment (bbrto21) [reply tools/run_command.py]: I renamed this

File: tools/run_command.py (hunk 17 above)
Comment (bbrto21) [reply tools/run_command.py]: I renamed this

File: tools/run_command.py [hunk 23]
```diff
@@ -29,13 +29,21 @@ def run_print_plugins(arv):
 
 
 commands = {}
-commands["tidy"] = run_check_tidy
-commands["test"] = run_integration_test
-commands["build"] = run_build_examples
-commands["plugins"] = run_print_plugins
+commands["tidy"] = {"func": run_check_tidy, "info": "Check and Updat format for C++ files"}
```
Comment (HakkyuKim) [inline tools/run_command.py:32]:
```suggestion
commands["tidy"] = {"func": run_check_tidy, "info": "Check and update format for C++ files"}
```

**Review by HakkyuKim — COMMENTED** [review]
Thanks for the changes, could we follow this style guide?

> Be consistent with your choice of string quote character within a file. Pick ' or " and stick with it. It is okay to use the other quote character on a string to avoid the need to \\ escape within the string.

https://google.github.io/styleguide/pyguide.html#310-strings

Plus, no extra spaces in f-string closure: `{ command }` -> `{command}`

File: tools/commands/build_example.py [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+
+
+def parse_args(args):
+    parser = command_utils.get_options_parser(
+        plugins=True, exclude=True, run_on_changed_packages=True, base_sha=True, command='build')
+    return parser.parse_args(args)
+
+
+def _build_examples(plugin):
+    name = plugin[plugin.rfind('/')+1:]
```
Comment (HakkyuKim) [inline tools/commands/build_example.py:17]: There is a convenient api:

```suggestion
    name = os.path.basename(plugin)
```

**Comment by bbrto21** [comment]
@HakkyuKim Thank you for your comments, and I applied your suggestions. please review again

**Review by HakkyuKim — APPROVED** [review]
Thanks for applying the changes. LGTM!

**Pull request merged**